### ISSUE TITLE
downgrade mkl to 2019.4 due to confirmed bug

### ISSUE THREAD

**Issue by stites** [NixOS/nixpkgs#75611]
**Describe the bug**
When running pytorch-1.3.1 tests with MKL, I get identical errors to https://github.com/hszhao/semseg/issues/16 and https://github.com/ContinuumIO/anaconda-issues/issues/11294.

According to the anaconda issue, this is a confirmed bug by the Intel OpenMP team and the workaround is to apply `export KMP_INIT_AT_FORK=FALSE` when running mkl code.

As NixOS is a general-purpose package manager, I think we should simply downgrade mkl to 2019.4 until further notice so that folks avoid the frustration for this "to-be-patched" version of mkl.

**To Reproduce**:
To reproduce, I expect that setting `doCheck=true` on the current pytorch derivation will replicate this issue. Alternatively, I placed pytorch in a shell.nix and ran the the tests directly in the v1.3.1 `test/` folder of the pytorch repo:

```
git clone https://github.com/pytorch/pytorch.git
git clone https://github.com/stites/pytorch-world.git
# cd pytorch-world && nix-shell ./release.nix -A pytorchWithMkl.env
# Actually pytorch-world is pinned so you'll have to bump pin/nixpkgs.json
cd ../pytorch && git fetch --tags && git checkout v1.2.0 # it is assumed to break, technically everything is on v1.3.1
$ python run_test.py
```

**Expected behavior**
Test suite runs

**Result**
The output of `python run_test.py` is an I/O abort signal:
```
Test executor: ['/nix/store/4kwi3im6k9y0z07k2l3ldhd1vmxl3vaf-python3-3.7.5-env/bin/python3.7']
Running test_autograd ... [2019-12-13 13:37:41.250160]
OMP: Error #13: Assertion failure at z_Linux_util.cpp(2361).
OMP: Hint Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see http://www.intel.com/software/products/support/.
Traceback (most recent call last):
  File "run_test.py", line 440, in <module>
    main()
  File "run_test.py", line 432, in main
    raise RuntimeError(message)
RuntimeError: test_autograd failed! Received signal: SIGIOT
```

**Additional context**
I'm currently upgrading pytorch to 1.3.1 and will either submit patches to downgrade MKL or apply this magical export flag to fix the pytorch test suite depending on how conversation turns out here. I assume, however, that this affects all runtime dependencies of mkl.

**Metadata**
- system: `"x86_64-linux"`
- host os: `Linux 4.19.84, NixOS, 19.09.1300.08e503bac6d (Loris)`
- multi-user?: `yes`
- sandbox: `yes`
- version: `nix-env (Nix) 2.3`
- channels(stites): `"home-manager"`
- channels(root): `"nixos-19.09.1300.08e503bac6d, unstable-20.03pre202154.58fb23f72ad"`
- nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
attribute:
- mkl
- pytorch
- numpy
```

**Comment by veprbl** [comment]
cc @smaret @markuskowa @bhipple

**Comment by markuskowa** [comment]
This should only be a problem on the unstable channel. On release-19.09 we have mkl-2019.3, which should not have that problem? On the intel home page there is mkl 2020 mentioned (https://software.intel.com/en-us/articles/intel-math-kernel-library-release-notes-and-new-features). I think we should try with upgrading first.

**Comment by bhipple** [comment]
I'd prefer to upgrade to MKL 2020, if it fixes the bug in question.  That said, a cursory glance at the release notes and bugfix log for 2019.5 and 2020.1 doesn't appear to mention this issue

If MKL 2020 doesn't fix the issue, I'd prefer a downgrade to the stable 2019.3 rather than patching the magic env var into every package using MKL.

**Comment by stites** [comment]
Happy to punt this, I've added the workaround [to the pytorch tests, here](https://github.com/NixOS/nixpkgs/pull/75827/files#diff-84b29016f8c93846ea69f3885d709770R235).

I guess the release cycle is long enough that we can simply wait (`unstable` only offers up unstable packages, after all). It looks like someone just confirmed today that [the download link has been fixed](https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/841834). I'll try to convert this issue into an update request.
